semanticdb-javac/src/main/java/com/sourcegraph/semanticdb_javac/SemanticdbVisitor.java

```diff
@@ -364,78 +467,108 @@ private String semanticdbText() {
     properties |= abstractNotDefault ? Property.ABSTRACT_VALUE : 0;
-    properties |= (sym.flags() & Flags.FINAL) > 0 ? Property.FINAL_VALUE : 0;
-    properties |= (sym.flags() & Flags.DEFAULT) > 0 ? Property.DEFAULT_VALUE : 0;
+    properties |= modifiers.contains(Modifier.FINAL) ? Property.FINAL_VALUE : 0;
+    properties |= isDefault ? Property.DEFAULT_VALUE : 0;
     return properties;
   }
 
-  private List<String> semanticdbParentSymbols(Symbol sym, List<String> result) {
-    if (!(sym instanceof Symbol.ClassSymbol)) {
-      return result;
-    }
-    Symbol.ClassSymbol csym = (Symbol.ClassSymbol) sym;
-    if (csym.getSuperclass() != Type.noType) {
-      semanticdbParentSymbol(csym.getSuperclass().tsym, result);
+  private List<String> semanticdbParentSymbols(TypeElement typeElement) {
+    ArrayList<String> parentSymbols = new ArrayList<>();
+    Set<TypeElement> parentElements = semanticdbParentTypeElements(typeElement, new HashSet<>());
+    for (TypeElement parentElement : parentElements) {
+      String ssym = semanticdbSymbol(parentElement);
+      if (!Objects.equals(ssym, SemanticdbSymbols.NONE)) {
+        parentSymbols.add(ssym);
+      }
     }
-    for (Type iType : csym.getInterfaces()) {
-      semanticdbParentSymbol(iType.tsym, result);
+    return parentSymbols;
+  }
+
+  private Set<TypeElement> semanticdbParentTypeElements(
+      TypeElement typeElement, Set<TypeElement> result) {
+    TypeMirror superType = typeElement.getSuperclass();
+    semanticdbParentSymbol(superType, result);
+    for (TypeMirror interfaceType : typeElement.getInterfaces()) {
+      semanticdbParentSymbol(interfaceType, result);
     }
+
     return result;
   }
 
-  private void semanticdbParentSymbol(Symbol sym, List<String> result) {
-    if (sym == null) {
-      return;
-    }
-    String ssym = semanticdbSymbol(sym);
-    if (!Objects.equals(ssym, SemanticdbSymbols.NONE)) {
-      result.add(ssym);
-      semanticdbParentSymbols(sym, result);
+  private void semanticdbParentSymbol(TypeMirror elementType, Set<TypeElement> result) {
+    if (!(elementType instanceof NoType)) {
+      Element superElement = types.asElement(elementType);
+      if (superElement != null && superElement instanceof TypeElement) {
+        result.add((TypeElement) superElement);
+        semanticdbParentTypeElements((TypeElement) superElement, result);
+      }
     }
   }
 
-  private List<String> semanticdbOverrides(Symbol sym) {
-    ArrayList<String> overriddenSymbols = new ArrayList<>();
-    Set<Symbol.MethodSymbol> overriddenMethods = javacTypes.getOverriddenMethods(sym);
-
-    for (Symbol.MethodSymbol meth : overriddenMethods) {
-      overriddenSymbols.add(semanticdbSymbol(meth));
+  private Set<String> semanticdbOverrides(
+      ExecutableElement sym, Element enclosingElement, HashSet<String> overriddenSymbols) {
+    if (enclosingElement instanceof TypeElement) {
```

Comment:
This change is my biggest concern about changing this PR: 1) it might be a big performance regression and 2) it might cause behavioral changes. What do you think about invoking `getOverriddenMethods` via reflection with a fallback to this implementation? Alternatively, we use `getOverriddenMethods` directly and open a separate PR that only refactors this implementation to use a non-public API.

I'm eager to merge this PR as soon as possible since it contains so many important improvements. Using `getOverriddenMethods` for now is not going to eliminate all uses of private APIs for now, but it's going to be much easier to fix this single case once we have merged everything else.